FastSimulation/CTPPSFastTrackingProducer/plugins/CTPPSFastTrackingProducer.cc

```diff
@@ -236,40 +221,46 @@ bool CTPPSFastTrackingProducer::SearchTrack(int i,int j,int Direction,double& xi
         station = &*pps_stationB;
     }
     if (det1->ppsNHits_<=i||det2->ppsNHits_<=j) return false;
+//
     double x1 = det1->ppsX_.at(i); double y1 = det1->ppsY_.at(i);
     double x2 = det2->ppsX_.at(j); double y2 = det2->ppsY_.at(j);
     double eloss;
+
     //thx and thy are returned in microrad
-    ReconstructArm(station,x1,y1,x2,y2,thx,thy,eloss);
+    ReconstructArm(station,Direction*x1,y1,Direction*x2,y2,thx,thy,eloss); // Pass the hits in the LHC ref. frame
+    thx*=-Direction;//  invert to the CMS ref frame
+    //thy*=-1;
+
     // Protect for unphysical results
     if (std::isnan(eloss)||std::isinf(eloss)||
             std::isnan(thx)  || std::isinf(thx) ||
             std::isnan(thy)  || std::isinf(thy)) return false;
     //
-    if (-thx<-100||-thx>300) return false;
-    if (thy<-200||thy>200) return false;
-    //
-    if ( m_verbosity ) std::cout << "thx " << thx << " thy " << thy << " eloss " << eloss << std::endl;
+    //if (-thx<-100||-thx>300) return false;
+    //if (thy<-200||thy>200) return false;
+    
+    if ( m_verbosity ) LogDebug("CTPPSFastTrackingProducer::SearchTrack:") << "thx " << thx << " thy " << thy << " eloss " << eloss;
 
-    // Get the start point of the reconstructed track near the origin made by Hector
-    x0 = -station->getX0()*um_to_cm;
+    // Get the start point of the reconstructed track near the origin made by Hector in the CMS ref. frame
+    x0 = -Direction*station->getX0()*um_to_cm;
     y0 = station->getY0()*um_to_cm;
     double ImpPar=sqrt(x0*x0+y0*y0);
     if (ImpPar>fImpParcut) return false;
     if (eloss<0.||eloss>fBeamEnergy) return false;
+//
     // Calculate the reconstructed track parameters 
     theta = sqrt(thx*thx+thy*thy)*urad;
     xi    = eloss/fBeamEnergy;
     double energy= fBeamEnergy*(1.-xi);
-    partP = sqrt(energy*energy-ProtonMassSQ);
-    t     = -2.*(ProtonMassSQ - fBeamEnergy*energy + fBeamMomentum*partP*cos(theta));
+    partP = sqrt(energy*energy-pow(CLHEP::proton_mass_c2/GeV,2));
```

Comment:
@mundim , pow(CLHEP::proton_mass_c2/GeV,2) seems to be ineffective, why not ProtonMassSQ? This ProtonMassSQ may be static const or constexpr

Reply:
done.